service-workers/service-worker/fetch-event-within-sw-manual.html

```diff
@@ -0,0 +1,122 @@
+
+    nextWorker.addEventListener('statechange', () => {
+      if (nextWorker.state == 'redundant') {
+        reject(Error(`Service worker failed to install`));
+        return;
+      }
+      if (nextWorker.state == 'activated') {
+        resolve();
+      }
+    });
+  });
+}
+
+function closeAllNotifications() {
+
+}
+
+function registerSwAndOpenFrame() {
+  return reset().then(() => navigator.serviceWorker.register(worker, {scope: 'resources/'}))
+    .then(reg => regReady(reg))
+    .then(() => with_iframe('resources/simple.html'));
+}
+
+promise_test(() => {
+  return registerSwAndOpenFrame().then(iframe => {
+    return Promise.all([
+      iframe.contentWindow.fetch('dummy.txt').then(r => r.text()),
+      iframe.contentWindow.caches.open('test')
+        .then(cache =>
+          cache.add('dummy.txt').then(() => cache.match('dummy.txt'))
+        ).then(response => {
+          if (!response) return 'cache match failed';
+          return response.text();
+        })
+    ])
+  }).then(([fetchText, cacheText]) => {
+    assert_equals(fetchText, 'intercepted', 'fetch intercepted');
+    assert_equals(cacheText, 'intercepted', 'cache.add intercepted');
+  });
+}, 'Service worker intercepts requests from window');
+
+promise_test(() => {
+  return registerSwAndOpenFrame().then(iframe => {
+    return Promise.all([
+      iframe.contentWindow.fetch('dummy.txt-inner-fetch').then(r => r.text()),
+      iframe.contentWindow.fetch('dummy.txt-inner-cache').then(r => r.text())
+    ])
+  }).then(([fetchText, cacheText]) => {
+    assert_equals(fetchText, 'Hello world\n', 'fetch within SW not intercepted');
+    assert_equals(cacheText, 'Hello world\n', 'cache.add within SW not intercepted');
+  });
+}, `Service worker does not intercept fetch/cache requests within service worker`);
+
+promise_test(() => {
+  return Notification.requestPermission().then(permission => {
+    if (permission != "granted") {
+      throw Error('You must allow notifications for this origin before running this test.');
+    }
+    return registerSwAndOpenFrame();
+  }).then(iframe => {
+    return Promise.race([
+      new Promise(resolve => {
+        const bc = new BroadcastChannel('icon-request');
+        bc.onmessage = () => {
+          bc.close();
+          resolve();
+        };
+
+        const notification = new iframe.contentWindow.Notification('test', {
+          icon: 'notification-icon.png'
+        });
+        notification.close();
+      }),
+      wait(1000).then(() => { throw Error(`Did not capture icon request for notification created from page`); })
+    ]).then(() => Promise.race([
+      new Promise(resolve => {
+        const bc = new BroadcastChannel('icon-request');
+        bc.onmessage = () => {
+          bc.close();
+          resolve();
+        };
+
+        iframe.contentWindow.fetch('show-notification');
+      }),
+      wait(1000).then(() => { throw Error(`Did not capture icon request for notification created within SW`); })
```

Comment:
Interesting. Once the notification appears in the sequence returned by `registration.getNotifications()`, it must have made the fetch for the icon (if the browser is following [the spec](https://notifications.spec.whatwg.org/#display-steps)). But that means there's a race between calling `showNotification` and `getNotification`, since `showNotification` resolves before icon fetches.